python/rascaline/rascaline/utils/clebsch_gordan/_cartesian_spherical.py

```diff
@@ -0,0 +1,416 @@
+
+    if cg_backend is None:
+        # TODO: benchmark & change the default?
+        if use_torch:
+            cg_backend = "python-dense"
+        else:
+            cg_backend = "python-sparse"
+
+    new_component_names: List[str] = []
+    for idx in range(len(tensor.component_names)):
+        if idx + 1 in axes_to_convert:
+            if len(axes_to_convert) == 1:
+                new_component_names.append("o3_mu")
+            else:
+                new_component_names.append(f"__internal_to_convert_{idx}")
+        else:
+            new_component_names.append("")
+
+    # Step 1: transform xyz dimensions to o3_lambda=1 dimensions
+    # This is done with `roll`, since (y, z, x) is the same as m = (-1, 0, 1)
+    new_blocks: List[TensorBlock] = []
+    for block in tensor.blocks():
+        values = _dispatch.roll(
+            block.values,
+            shifts=[-1] * len(axes_to_convert),
+            axis=axes_to_convert,
+        )
+
+        new_components: List[Labels] = []
+        for idx, component in enumerate(block.components):
+            if idx + 1 in axes_to_convert:
+                new_components.append(
+                    Labels(
+                        names=[new_component_names[idx]],
+                        values=_dispatch.int_array_like(
+                            [[-1], [0], [1]], component.values
+                        ),
+                    )
+                )
+            else:
+                new_components.append(component)
+
+        new_blocks.append(
+            TensorBlock(values, block.samples, new_components, block.properties)
+        )
+
+    if tensor_rank == 1:
+        new_keys = tensor.keys
+        # we are done, add o3_lambda/o3_sigma/l_1 to the keys & return
+        if keep_l_in_keys:
+            new_keys = new_keys.insert(
+                0,
+                "l_1",
+                _dispatch.int_array_like([1] * len(new_keys), new_keys.values),
+            )
+
+        new_keys = new_keys.insert(
+            0,
+            "o3_sigma",
+            _dispatch.int_array_like([1] * len(tensor.keys), tensor.keys.values),
+        )
+
+        new_keys = new_keys.insert(
+            0,
+            "o3_lambda",
+            _dispatch.int_array_like([1] * len(tensor.keys), tensor.keys.values),
+        )
+
+        return TensorMap(new_keys, new_blocks)
+
+    # Step 2: if there is more than one dimension, couple them with CG coefficients
+    if cg_coefficients is None:
+        if torch_jit_is_scripting():
+            raise ValueError(
+                "in TorchScript mode, `cg_coefficients` must be pre-computed "
+                "and given to this function explicitly"
+            )
+        else:
+            cg_coefficients = _coefficients.calculate_cg_coefficients(
+                lambda_max=len(axes_to_convert),
+                cg_backend=cg_backend,
+                use_torch=use_torch,
+            )
+
+    iteration_index = 0
+    while len(axes_to_convert) > 1:
```

Comment:
Original Q: Doesn't this ignore the final axis?

Discussed - no because this is iterating over combination pairs, leaving the last axis where it is.